test/integration_test.py

```diff
@@ -214,36 +211,101 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
-
-        sources = sorted(self.azul_client.catalog_sources(catalog))
-        # See below why we shuffle
-        self.random.shuffle(sources)
-        bundle_fqids = chain.from_iterable(
-            self.azul_client.list_bundles(catalog, source, prefix(source))
-            for source in sources
-        )
-        log.info('Randomly selecting %i bundles from catalog %s.', max_bundles, catalog)
-        # No point in exhausting the iterator once we have a reasonable amount
-        # of inputs to randomly select from. This truncation prefers sources
-        # occurring first, so we shuffle them above to neutralize the bias.
-        bundle_fqids = islice(bundle_fqids, max_bundles * 10)
-        bundle_fqids = reservoir_sample(max_bundles, bundle_fqids, random=self.random)
-
-        if len(bundle_fqids) >= max_bundles:
-            log.warning('Not enough bundles in catalog %r. The test may fail.', catalog)
-
-        return bundle_fqids
+        sources = iter(self.random.sample(plugin.sources, len(plugin.sources)))
+        while total_bundles < min_bundles:
+            try:
+                source = next(sources)
+            except StopIteration:
+                log.warning('Checked all sources and found only %d bundles instead of the '
+                            'expected minimum %d', total_bundles, min_bundles)
+                break
+            source = plugin.resolve_source(str(source))
+            prefix, new_fqids = self._choose_prefix(catalog, source)
+            total_bundles += len(new_fqids)
+            yield source, prefix, new_fqids
+
+    def _list_managed_access_bundles(self,
+                                     catalog: CatalogName
+                                     ) -> Iterator[Tuple[SourceRef, str, List[SourcedBundleFQID]]]:
+        sources = self.azul_client.catalog_sources(catalog)
+        # We need at least one managed_access bundle per IT. To index them with
+        # remote_reindex and avoid collateral bundles, we use as specific a
+        # a prefix as possible.
+        for source in self.managed_access_sources_by_catalog[catalog]:
+            assert str(source.spec) in sources
+            bundle_fqid = self.random.choice(
+                self.azul_client.list_bundles(catalog, source, prefix='')
+            )
+            # FIXME: We shouldn't need to include the common prefix
+            #        https://github.com/DataBiosphere/azul/issues/3579
+            common = source.spec.effective.prefix.common
+            prefix = bundle_fqid.uuid[len(common):8]
+            assert prefix != '', prefix
+            new_fqids = self.azul_client.list_bundles(catalog, source, prefix)
+            yield source, prefix, new_fqids
+
+    def _choose_prefix(self,
+                       catalog: CatalogName,
+                       source: SourceRef
+                       ) -> Tuple[str, List[SourcedBundleFQID]]:
+        fqids = self.azul_client.list_bundles(catalog, source, '')
```

Comment:
This looks like it lists all bundles in the source. If that's true, this approach would be infeasible, as I mentioned in a previous parking lot discussion with you. If we really need to count bundles, we need to add a plugin method for just that.

But: I remember giving different directives to the previous author. We currently (on develop) hard-code the prefix and the partition prefix length in the configuration (with the help of a heuristic) to get a reasonable number of bundles per partition. We decided to add a test to check the hardcoded prefix to make sure it doesn't produce an empty source. We should extend that check to ensure it doesn't produce overly small or overly large partitions. Once we have that, we simply pick a random partition from a set of randomly selected sources and we're done.